File: DesignPatterns/Duck/Duck.cpp
#include "Duck.h"

Duck::Duck()
{
}


Duck::~Duck()
{
}
void Duck::display()
{
	std::cout << "A Duck" << std::endl;
}
void Duck::performFly()
{
	flyBehavior.fly();
}
void Duck::performQuack()
{
	quackBehavior.quack();
}
void Duck::swim()
{
	std::cout << "All Duck floats." << std::endl;
}

File: my-app/src/main/java/com/mycompany/app/Duck.java
package com.mycompany.app;

public abstract class Duck{
    QuackBehavior quackBehavior;
    FlyBehavior flyBehavior;
    public Duck(){

    }
    public void performFly(){
        flyBehavior.fly();
    }
    private interface QuackBehavior{
        public void quack();
    }
    private interface FlyBehavior{
        public void fly();
    }
    
    private class Quack implements QuackBehavior{
        public void quack() {
            System.out.println("Quack");
        }
    }
    private class MuteQuack implements QuackBehavior{
        public void quack() {
            System.out.println("Silence");
        }
    }
}
